### PR TITLE
gh-120496: Use CAS approach for rangeiter_next

### PR DESCRIPTION
**Pull request by corona10**
* Issue: gh-120496

### CODE REVIEW

**Comment by Fidget-Spinner** [comment]
Thanks for doing this! Do you have any benchmark results on how much the perf impact is? versus say using `Py_BEGIN_CRITICAL_SECTION` ?

**Comment by corona10** [comment]
>Thanks for doing this! Do you have any benchmark results on how much the perf impact is? versus say using 

I will try to prepare :)

**Comment by corona10** [comment]
@Fidget-Spinner @colesbury 

I am not sure this will be a fair benchmark, but it shows different strengths in terms of technique.
Single-thread: critical section win
Multi-thread: CAS win

## Single thread
```python
import pyperf

runner = pyperf.Runner()
runner.timeit(name="bench iter single",
              stmt="""
data = list(it)
""",
              setup = """
N = 10
it = iter(range(N))
"""
              )
```
* Mean +- std dev: [single_baseline] 80.0 ns +- 2.3 ns -> [single_cas] 82.9 ns +- 2.0 ns: 1.04x slower

## Mutl thread
```

import pyperf

runner = pyperf.Runner()
runner.timeit(name="bench iter multi",
              stmt="""
for _ in range(100):
    it = iter(range(N))
    with concurrent.futures.ThreadPoolExecutor() as executor:
        data = list(executor.map(lambda _: next(it), range(N)))
""",
              setup = """
import concurrent.futures
N = 100
"""
              )
```
* Mean +- std dev: [multi_baseline] 284 ms +- 11 ms -> [multi_cas] 277 ms +- 20 ms: 1.03x faster

**Comment by Fidget-Spinner** [comment]
Hmmm the perf difference is not that large. That might imply that we should just use critical section for simplicity, and hope that someone on PyPI publishes a "prange()" or something like that.

**Comment by corona10** [comment]
>  That might imply that we should just use critical section for simplicity, and hope that someone on PyPI publishes a "prange()" or something like that.

Yeah, I prefer this one.

File: Objects/rangeobject.c [hunk 1]
```diff
@@ -816,13 +816,29 @@ PyTypeObject PyRange_Type = {
...
 {
+#ifdef Py_GIL_DISABLED
+    // r->step is readonly attribute, so we can assume it is not changed.
+    long step = r->step;
+    do {
+        long len = _Py_atomic_load_long_relaxed(&r->len);
+        if (len <= 0) {
+            return NULL;
+        }
+        long result = _Py_atomic_load_long_relaxed(&r->start);
```
Comment (eendebakpt) [inline Objects/rangeobject.c:827]: What if we have a range object with `len=1` and many threads simultaneously at this point? The check on `len` has already been done, so the threads each increment the value of `start` and we end up with a value for `result` larger than `end`.

Comment (corona10) [reply Objects/rangeobject.c:827]: Maybe fail at _Py_atomic_compare_exchange_long?

Comment (corona10) [reply Objects/rangeobject.c:827]: Ah.. yes understood what you pointed out.

Comment (corona10) [reply Objects/rangeobject.c:827]: There will be a slight timing issue. Then yes, let's just use the critical section that @Fidget-Spinner also preferred.

**Comment by corona10** [comment]
See: https://github.com/python/cpython/pull/120540

File: Objects/rangeobject.c (hunk 1 above)
Comment (corona10) [reply Objects/rangeobject.c:827]: To prepare for the overcommitted issue, we can validate the value of the result by comparing the end value.
But let's just use critical section approach.

**Pull request closed without merging**